server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,205 @@
+
+#[proxy(
+    default_service = "org.gnome.keyring.SystemPrompter",
+    interface = "org.gnome.keyring.internal.Prompter",
+    default_path = "/org/gnome/keyring/Prompter"
+)]
+pub trait Prompter {
+    fn begin_prompting(&self, callback: &OwnedObjectPath) -> Result<(), ServiceError>;
+
+    fn perform_prompt(
+        &self,
+        callback: OwnedObjectPath,
+        type_: &str,
+        properties: HashMap<&str, OwnedValue>,
+        exchange: &str,
+    ) -> Result<(), ServiceError>;
+
+    fn stop_prompting(&self, callback: OwnedObjectPath) -> Result<(), ServiceError>;
+}
+
+// org.gnome.keyring.internal.Prompter.Callback
+
+pub struct PrompterCallback {
+    service: Service,
+    path: OwnedObjectPath,
+}
+
+#[interface(name = "org.gnome.keyring.internal.Prompter.Callback")]
+impl PrompterCallback {
+    pub async fn prompt_ready(
+        &self,
+        reply: &str,
+        properties: HashMap<&str, OwnedValue>,
+        exchange: &str,
+        #[zbus(connection)] connection: &zbus::Connection,
+    ) -> Result<(), ServiceError> {
+        let Some(prompt) = self.service.prompt().await else {
+            return Err(ServiceError::NoSuchObject(
+                "Prompt does not exist.".to_string(),
+            ));
+        };
+
+        match prompt.role() {
+            PromptRole::Lock => {
+                if properties.is_empty() {
+                    // First PromptReady call
+                    let mut properties: HashMap<&str, zvariant::OwnedValue> = HashMap::new();
+                    let secret_exchange = SecretExchange::new().map_err(|err| {
+                        ServiceError::ZBus(zbus::Error::FDO(Box::new(zbus::fdo::Error::Failed(
+                            format!("Failed to generate SecretExchange {err}."),
+                        ))))
+                    })?;
+                    let exchange = secret_exchange.begin();
+
+                    properties.insert("continue-label", Value::new("Lock").try_to_owned().unwrap());
+                    properties.insert(
+                        "description",
+                        Value::new("Confirm locking \"login\", Keyring.")
+                            .try_to_owned()
+                            .unwrap(),
+                    );
+                    // TODO: set the object name dynamically in description
+                    properties.insert(
+                        "message",
+                        Value::new("Lock Keyring").try_to_owned().unwrap(),
+                    );
+                    properties.insert("caller-window", Value::new("").try_to_owned().unwrap());
+                    properties.insert("cancel-label", Value::new("Cancel").try_to_owned().unwrap());
+
+                    let connection = connection.clone();
+                    let path = self.path.clone();
+
+                    tokio::spawn(PrompterCallback::perform_prompt(
+                        connection, path, "confirm", properties, exchange,
+                    ));
```

Comment:
Not sure why this has to be in PrompterCallback,

Reply:
What do you mean? Maybe the confusion is caused by the fact that I'm using the same name  as `Prompter.perform_prompt`. `PrompterCallback::perform_prompt` is just a wrapper calling `Prompter.perform_prompt` with other code to avoid code duplication.